feat(appendTo): new parameter

Jerskouille opened the pull request:
**Summary**

See #135
New parameter: `appendToBody`.
Allows to not wrap the input in `.algolia-autocomplete`.

Automatically repositions itself on resize.
Limitation: requires `hint: false`. Indeed, we're not wrapping the input anymore, so we can't have it behind the input.

**Result**
Without much surprise, it looks the same way:
![Autocomplete with appendToBody](https://cloud.githubusercontent.com/assets/5095856/21457054/98380df8-c92c-11e6-8171-8c5a25f28565.png)

However, on the playground, it ends up "breaking" the example in `playground.html`.
But this is actually expected based on the bootstrap layout we're using, it's supposed to fail.

```html
<div class="input-group">
  <input class="form-control" />
  <input class="form-control" />
  <span class="input-group-addon">Go</span>
</div>
```

![image](https://cloud.githubusercontent.com/assets/5095856/21457188/80e91df8-c92d-11e6-8c44-bbb9a6103a49.png)

rayrutjes commented:
What do you think of implementing the suggestion of @vvo ?
https://github.com/algolia/autocomplete.js/issues/135#issuecomment-269003450

I suppose it could make sense to be able to attach it at given location, for styling purposes for example.

Awesome you started implementing this!

Jerskouille commented:
`appendTo` done, what are your thoughts @vvo @rayrutjes ?

Jerskouille commented:
I'm a bit unsure on how to test this parameter. In the end, we'd like to rerun basically the whole test suite (`hint` tests excluded) with `appendTo` I guess ?

redox commented:
Let me know if you need any help @Jerskouille 👍

Jerskouille commented:
Rebased and squashed, ready for merge.

Wrap-up:
- `appendTo`: Takes a selector, and will append to the first item matching the selector
- Runs the whole integration test suite once as before and once with `appendTo`
- Added an integration test to check the dropdown position (with a +/-2px tolerance for Safari (don't ask me why))
- The integration tests being, to me, enough, I've just added a single unit test to check that we `throw` when `appendTo` is used with `hint`.

Jerskouille commented:
cc @vvo @rayrutjes @redox

redox commented on src/autocomplete/typeahead.js:
```diff
@@ -31,12 +31,22 @@ function Typeahead(o) {
+  o.hint = !!o.hint;
+
+  if (o.hint && o.appendTo) {
+    throw new Error('[autocomplete.js] hint and appendTo options can\'t be used at the same time');
```
Why do we need to ensure that?

vvo replied:
I guess this is because when you use appendTo you are no more wrapping the input as @Jerskouille said.

Jerskouille replied:
^ What he said, clarified a bit more in my last comment (`appendTo` vs `dropdownMenuContainer`).

vvo commented:
Can you add the README documentation on this new feature? With an explanation on:
- when to/why use it
- effects and limitations (hint)

vvo commented on test/unit/typeahead_spec.js:
```diff
@@ -40,6 +40,20 @@ describe('Typeahead', function() {
   });
 
+
+  describe('appendTo', function() {
```
Did you try to add a unit test that checks that something is added to a particular node?

```js
var node = document.createElement('div');
document.querySelector('body').appendChild(node);

// instantiate with appendTo..
// test node now has children?..
```

Jerskouille replied:
Done here: https://github.com/algolia/autocomplete.js/pull/137/commits/b9e0afd17080561977a1daee418df65af19cdf7e#diff-dfaaccd0a4dbcd8f48b32d89fa8c52fcR56

vvo commented:
@Jerskouille Do you have any idea on how dropdownMenuContainer and appendTo are different now? Is the difference only about not wrapping the input completely? They seems very similar and yet I would not be able to tell which one does what right now. If you can help clearing this that would be nice.

Jerskouille commented:
> Do you have any idea on how dropdownMenuContainer and appendTo are different now? Is the difference only about not wrapping the input completely? They seems very similar and yet I would not be able to tell which one does what right now. If you can help clearing this that would be nice.

I've never used `dropdownMenuContainer` and am not sure why it was created in the first place.
On my side, the issues I had were not so much about the dropdown than the wrapping `span.algolia-autocomplete`.

To clear things up, to my understanding:

- `dropdownMenuContainer`: 
   - Removes the menu from the wrapping element and appends it somewhere else
   - also ensures the containing  element is `position: relative`
   - sets the menu to `top: 0`

- `appendTo`:
   - Never wraps the `input`
   - Takes `.algolia-autocomplete` and its content, and puts it elsewhere
   - The `hint` is disabled right now, because we're not wrapping anymore. We might be able to handle it, but z-index and positioning issues would arise in that case, and I wasn't up to it
   - Takes the full width of the associated input
   - Repositions on every keystroke under the input (which triggers a `redrawn` event in case you want to adjust it)

vvo commented:
Thanks a lot for clarification, should both option be doable at the same time? If not then we should throw also no?

Jerskouille commented:
> Thanks a lot for clarification, should both option be doable at the same time? If not then we should throw also no?

Good question, I hadn't thought about that. I think it will work, it just doesn't make much sense, because if you extract the wrapper AND the menu, re-extracting the menu from the wrapper would be weird. Do you want me to confirm?

vvo commented:
> Do you want me to confirm?

I think we are good, I do not expect that to be done by the user I guess,

Jerskouille commented on src/autocomplete/css.js:
```diff
@@ -61,6 +61,18 @@ var css = {
     dataset: 'dataset',
     empty: 'empty'
+  },
+  appendToBody: {
```
Done here: https://github.com/algolia/autocomplete.js/pull/137/commits/8d491f20af6a12f7d91ae6e42cb84a338d5398e1

Jerskouille commented:
The last changes should answer all the questions / worries you had. (README also updated)

redox commented:
Good for me 👍

vvo commented:
Let's merge and release early next week so we can follow the release.

vvo commented:
Thanks. A LOT! This ships today
